atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepMetrics.java

```diff
@@ -15,62 +15,106 @@
+    @VisibleForTesting
+    static final String ITERATION_TAG = "perIteration";
+    @VisibleForTesting
+    static final String TABLE_TAG = "perTable";
 
-    private class TableSpecificHistogramMetric {
-        private final String name;
+    private class SweepMetric {
+        private final String histogram;
+        private final String meter;
+        private final String gauge;
 
-        TableSpecificHistogramMetric(String name) {
-            this.name = name;
+        SweepMetric(String name) {
+            this.histogram = name + "H";
+            this.meter = name + "M";
+            this.gauge = name + "G";
         }
 
-        void update(long value) {
-            metricsManager.getRegistry().histogram(aggregateMetric()).update(value);
+        void update(long value, Optional<TableReference> tableRef) {
+            metricRegistry.histogram(getTaggedMetric(histogram, tableRef)).update(value);
+            metricRegistry.meter(getTaggedMetric(meter, tableRef)).mark(value);
         }
 
-        private String aggregateMetric() {
-            return MetricRegistry.name(SweepMetrics.class, name);
+        void updateTimeElapsed(long value) {
+            metricRegistry.histogram(getTaggedMetric(histogram, Optional.empty())).update(value);
+            ((Current) metricRegistry.gauge(getTaggedMetric(gauge, Optional.empty()), new Current())).setValue(value);
         }
+
     }
 
-    private class MeterMetric {
-        private final Meter meter;
+    @VisibleForTesting
+    static class Current implements Gauge<Long> {
+        private volatile long value = 0;
+
+        @Override
+        public Long getValue() {
+            return value;
+        }
 
-        MeterMetric(String name) {
-            this.meter = metricsManager.registerOrGetMeter(SweepMetrics.class, null, name);
+        public void setValue(long newValue) {
+            value = newValue;
         }
+    }
+
+    private MetricName getTaggedMetric(String name, Optional<TableReference> tableRef) {
+        return MetricName.builder()
+                .safeName(MetricRegistry.name(SweepMetrics.class, name))
+                .safeTags(tableRef.map(SweepMetrics::getTag).orElse(ImmutableMap.of(PERIOD_TAG, ITERATION_TAG)))
+                .build();
+    }
 
-        void update(long value) {
-            this.meter.mark(value);
+    private static Map<String, String> getTag(TableReference tableRef) {
+        Arg<String> maybeSafeTableRef = LoggingArgs.tableRef(tableRef);
+        if (maybeSafeTableRef.isSafeForLogging()) {
+            return ImmutableMap.of(maybeSafeTableRef.getName(), maybeSafeTableRef.getValue(), PERIOD_TAG, TABLE_TAG);
+        } else {
+            return ImmutableMap.of("unsafeTableRef", "unsafe", PERIOD_TAG, TABLE_TAG);
         }
     }
 
-    void examinedCells(long numExamined) {
-        cellsSweptHistogram.update(numExamined);
-        cellsSweptMeter.update(numExamined);
+    void updateMetricsOneIteration(SweepResults sweepResults) {
```

Comment:
Few thoughts on this class still, it isn't sitting right with me.  I think we should tag the `OneIteration` metrics with the table name as well - we have the tableRef at the call site and it will allow us to compare table1 vs table 2 or table1 at customer1 vs table1 at customer 2.

I'm also not sure it is possible to plot a metric where a tag isn't present in DD.  I can filter to a single tag value, but I can't say "give me these metric values where this tag isn't specified".

Secondly, I still think we should have separate metrics for incremental and end-of-table metrics.  These are updated at different rates, I wouldn't want to plot incremental and end-of-table as a single metric or compare them against each other.  They are designed to tell us different things and create different monitors.


There's a also 'code smell' in this class where we have the same code path duplicated.